README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

### *Attention!*

Google has [recently changed its API
requirements](https://developers.google.com/maps/documentation/geocoding/usage-and-billing),
and **ggmap** users are now required to register with Google. From a
user’s perspective, there are essentially three ramifications of this:

1.  Users must register with Google. You can do this at
    <a href="https://cloud.google.com/maps-platform/" class="uri">https://cloud.google.com/maps-platform/</a>.
    While it will require a valid credit card (sorry!), there seems to
    be a fair bit of free use before you incur charges, and even then
    the charges are modest for light use.

2.  Users must enable the APIs they intend to use. What may appear to
    **ggmap** users as one overarching “Google Maps” product, Google in
    fact has several services that it provides as geo-related solutions.
    For example, the [Maps Static
    API](https://developers.google.com/maps/documentation/maps-static/intro)
    provides map images, while the [Geocoding
    API](https://developers.google.com/maps/documentation/geocoding/intro)
    provides geocoding and reverse geocoding services. Apart from the
    relevant Terms of Service, generally **ggmap** users don’t need to
    think about the different services. For example, you just need to
    remember that `get_googlemap()` gets maps, `geocode()` geocodes
    (with Google, DSK is done), etc., and **ggmap** handles the queries
    for you. *However*, you do need to enable the APIs before you use
    them. You’ll only need to do that once, and then they’ll be ready
    for you to use. Enabling the APIs just means clicking a few radio
    buttons on the Google Maps Platform web interface listed above, so
    it’s easy.

3.  Inside R, after loading the new version of **ggmap**, you’ll need
    provide **ggmap** with your API key, a [hash
    value](https://en.wikipedia.org/wiki/Hash_function) (think string of
    jibberish) that authenticates you to Google’s servers. This can be
    done on a temporary basis with `register_google(key = "[your key]")`
    or permanently using
    `register_google(key = "[your key]", write = TRUE)` (note: this will
    overwrite your `~/.Renviron` file by replacing/adding the relevant
    line). If you use the former, know that you’ll need to re-do it
    every time you reset R.

Your API key is *private* and unique to you, so be careful not to share
it online, for example in a GitHub issue or saving it in a shared R
script file. If you share it inadvertantly, just get on Google’s website
and regenerate your key - this will retire the old one. Keeping your key
private is made a bit easier by **ggmap** scrubbing the key out of
queries by default, so when URLs are shown in your console, they’ll look
something like `key=xxx`. (Read the details section of the
`register_google()` documentation for a bit more info on this point.)

We hope the new version of **ggmap** will be on CRAN soon, but until
then you can install the version here with:

``` r
if(!requireNamespace("devtools")) install.packages("devtools")
devtools::install_github("dkahle/ggmap")
```

The details of the readme below will be changed shortly to reflect these
changes. Thanks for your patience!

<hr>

ggmap
=====

**ggmap** is an R package that makes it easy to retrieve raster map
tiles from popular online mapping services like [Google
Maps](https://developers.google.com/maps/documentation/static-maps/?hl=en)
and [Stamen Maps](http://maps.stamen.com) and plot them using the
[**ggplot2**](https://github.com/tidyverse/ggplot2) framework:

``` r
library("ggmap")

us <- c(left = -125, bottom = 25.75, right = -67, top = 49)
get_stamenmap(us, zoom = 5, maptype = "toner-lite") %>% ggmap() 
```

![](tools/README-maptypes-1.png)

Use `qmplot()` in the same way you’d use `qplot()`, but with a map
automatically added in the background:

``` r
library("dplyr")
library("forcats")

# define helper
`%notin%` <- function(lhs, rhs) !(lhs %in% rhs)

# reduce crime to violent crimes in downtown houston
violent_crimes <- crime %>% 
  filter(
    offense %notin% c("auto theft", "theft", "burglary"),
    -95.39681 <= lon & lon <= -95.34188,
     29.73631 <= lat & lat <=  29.78400
  ) %>% 
  mutate(
    offense = fct_drop(offense),
    offense = fct_relevel(offense, c("robbery", "aggravated assault", "rape", "murder"))
  )

# use qmplot to make a scatterplot on a map
qmplot(lon, lat, data = violent_crimes, maptype = "toner-lite", color = I("red"))
```

![](tools/README-qmplot-1.png)

All the **ggplot2** geom’s are available. For example, you can make a
contour plot with `geom = "density2d"`:

``` r
qmplot(lon, lat, data = violent_crimes, maptype = "toner-lite", geom = "density2d", color = I("red"))
```

In fact, since **ggmap**’s built on top of **ggplot2**, all your usual
**ggplot2** stuff (geoms, polishing, etc.) will work, and there are some
unique graphing perks **ggmap** brings to the table, too.

``` r
robberies <- violent_crimes %>% filter(offense == "robbery")

qmplot(lon, lat, data = violent_crimes, geom = "blank", 
  zoom = 14, maptype = "toner-background", darken = .7, legend = "topleft"
) +
  stat_density_2d(aes(fill = ..level..), geom = "polygon", alpha = .3, color = NA) +
  scale_fill_gradient2("Robbery\nPropensity", low = "white", mid = "yellow", high = "red", midpoint = 650)
```

![](tools/README-styling-1.png)

Faceting works, too:

``` r
qmplot(lon, lat, data = violent_crimes, maptype = "toner-background", color = offense) + 
  facet_wrap(~ offense)
```

![](tools/README-faceting-1.png)

Google Maps and Credentials
---------------------------

[Google Maps](http://developers.google.com/maps/terms) can be used just
as easily. However, since Google Maps use a center/zoom specification,
their input is a bit different:

``` r
get_googlemap("waco texas", zoom = 12) %>% ggmap()
#  Source : https://maps.googleapis.com/maps/api/staticmap?center=waco%20texas&zoom=12&size=640x640&scale=2&maptype=terrain&key=xxx
#  Source : https://maps.googleapis.com/maps/api/geocode/json?address=waco+texas&key=xxx
```

![](tools/README-google_maps-1.png)

Moreover, you can get various different styles of Google Maps with
**ggmap** (just like Stamen Maps):

``` r
get_googlemap("waco texas", zoom = 12, maptype = "satellite") %>% ggmap()
get_googlemap("waco texas", zoom = 12, maptype = "hybrid") %>% ggmap()
get_googlemap("waco texas", zoom = 12, maptype = "roadmap") %>% ggmap()
```

Google’s geocoding and reverse geocoding API’s are available through
`geocode()` and `revgeocode()`, respectively:

``` r
geocode("1301 S University Parks Dr, Waco, TX 76798")
#  Source : https://maps.googleapis.com/maps/api/geocode/json?address=1301+S+University+Parks+Dr,+Waco,+TX+76798&key=xxx
#  # A tibble: 1 x 2
#      lon   lat
#    <dbl> <dbl>
#  1 -97.1  31.6
revgeocode(c(lon = -97.1161, lat = 31.55098))

NEWS
ggmap 3.0.0
-------------------------------------------------------------------------

CHANGES

* In mid-2018, Google merged its maps-related online services into its
  cloud platform. This change broke many ggmap functions, and others had
  atrophied by simple changing in underlying web APIs.  As a consequence,
  CloudMade, OSM, and Naver are no longer available. Google services now
  require registering with Google, see ?register_google for details on
  how to get up and running with that. The service provides a good
  alotment of free use, but it is fundamentally a paid service.
* ggmap now makes widespread, though not yet exhaustive, use of tidyverse
  principles and functions. For example, most functions that return
  information (e.g. geocode) now return tibbles. Web queries are made
  with httr, which helps alleviate strange download problems that
  frustrated users in the past. Over time this dependence will likely
  become stronger.
* Mikko Korpela (@mvkorpel) has been added as a contributor.  Thanks for
  your additions, Mikko!
* Scott M M Jackson (@scottmmjackson) has been added as an author. Scott
  has been a huge help with maintenance and has some great ideas for the
  future. Thanks, Scott!



ggmap 2.7
-------------------------------------------------------------------------

NEW FEATURES

* stamen now provides its terrain tileset for the whole globe.
* all google functions can now use google credentials stored in
  options("ggmap") to better interface with those components of the google
  api. make sure your api key is registered for each before you use them.
  (#29, #92, #83)
* geocode() now has a force option to override cached values. (#69,
  thanks @samfield)
* mutate_geocode() now works for tibbles. (#106, thanks @voxnonecho)
* trek() now allows users to get routes from google that aren't straight
  lines on plots. (#99, thanks stackoverflow user @akhmed) this required
  importing bitops.
* google functions have domain name extension options, urlonly options,
  and inject options for any code that you manually want to pass to the
  url (i.e. anything that isn't built into ggmap).
* mode = "transit" can now be used with mapdist().


CHANGES

* has_goog*() functions are now has_google*(); the former are .Deprecated().
* dplyr is now listed as a dependency (for bind_cols in mutate_geocode).
* get_stamenmap() now issues a warning if a location parameter is passed
  (#55, thanks @cguwilliams).


FIXES

* get_stamenmap() now works again for terrain and terrain-background
  images (#102, thanks @UrbanDemog).
* a bug in parsing the base_layer argument is now fixed. (thanks to
  @restonslacker)
* gglocator now properly adapts to scaling. (minor update, thanks to
  @Nikolai-Hlubek)
* google functions now use enc2utf8() to better handle foreign character
  sets. (#98, #89, thanks dapper2dan)
* google functions now better handle special characters. (#104, thanks
  casa-henrym)
* gglocator now properly works again. (#87, thanks @salah31416 and @mvkorpel)
* README images moved from figures/ to tools/
* mutate_geocode() now works with tibbles (#68, thanks dannguyen)
* get_stamenmap() now properly changes colors to black and white.


ggmap 2.6.1
-------------------------------------------------------------------------

FIXES

* fixes a bug in qmplot that gives the error "Error in unlist(.all_aesthetics[1:42]) :
  object '.all_aesthetics' not found". (thanks to @Hughan.)

R/ggmap.R
#' Plot a ggmap object
#'
#' ggmap plots the raster object produced by \code{\link{get_map}}.
#'
#' @param ggmap an object of class ggmap (from function get_map)
#' @param extent how much of the plot should the map take up?
#'   "normal", "device", or "panel" (default)
#' @param base_layer a ggplot(aes(...), ...) call; see examples
#' @param maprange logical for use with base_layer; should the map
#'   define the x and y limits?
#' @param legend "left", "right" (default), "bottom", "top",
#'   "bottomleft", "bottomright", "topleft", "topright", "none"
#'   (used with extent = "device")
#' @param padding distance from legend to corner of the plot (used
#'   with legend, formerly b)
#' @param darken vector of the form c(number, color), where number
#'   is in [0, 1] and color is a character string indicating the
#'   color of the darken.  0 indicates no darkening, 1 indicates a
#'   black-out.
#' @param b Deprecated, renamed to `padding`. Overrides any
#'   `padding` argument.
#' @param fullpage Deprecated, equivalent to `extent = "device"`
#'   when `TRUE`. Overrides any `extent` argument.
#' @param expand Deprecated, equivalent to `extent = "panel"`
#'   when `TRUE` and `fullpage` is `FALSE`. When `fullpage`
#'   is `FALSE` and `expand` is `FALSE`, equivalent to
#'   `extent="normal"`. Overrides any `extent` argument.
#' @param ... ...
#' @return a ggplot object
#' @author David Kahle \email{david.kahle@@gmail.com}
#' @seealso \code{\link{get_map}}, \code{\link{qmap}}
#' @export ggmap inset inset_raster
#' @examples
#'
#' \dontrun{## map queries drag R CMD check
#'
#'
#' ## extents and legends
#' ##################################################
#' hdf <- get_map("houston, texas")
#' ggmap(hdf, extent = "normal")
#' ggmap(hdf) # extent = "panel", note qmap defaults to extent = "device"
#' ggmap(hdf, extent = "device")
#'
#'
#'
#' # make some fake spatial data
#' mu <- c(-95.3632715, 29.7632836); nDataSets <- sample(4:10,1)
#' chkpts <- NULL
#' for(k in 1:nDataSets){
#'   a <- rnorm(2); b <- rnorm(2);
#'   si <- 1/3000 * (outer(a,a) + outer(b,b))
#'   chkpts <- rbind(
#'     chkpts,
#'     cbind(MASS::mvrnorm(rpois(1,50), jitter(mu, .01), si), k)
#'   )
#' }
#' chkpts <- data.frame(chkpts)
#' names(chkpts) <- c("lon", "lat","class")
#' chkpts$class <- factor(chkpts$class)
#' qplot(lon, lat, data = chkpts, colour = class)
#'
#' # show it on the map
#' ggmap(hdf, extent = "normal") +
#'   geom_point(aes(x = lon, y = lat, colour = class), data = chkpts, alpha = .5)
#'
#' ggmap(hdf) +
#'   geom_point(aes(x = lon, y = lat, colour = class), data = chkpts, alpha = .5)
#'
#' ggmap(hdf, extent = "device") +
#'   geom_point(aes(x = lon, y = lat, colour = class), data = chkpts, alpha = .5)
#'
#' theme_set(theme_bw())
#' ggmap(hdf, extent = "device") +
#'   geom_point(aes(x = lon, y = lat, colour = class), data = chkpts, alpha = .5)
#'
#' ggmap(hdf, extent = "device", legend = "topleft") +
#'   geom_point(aes(x = lon, y = lat, colour = class), data = chkpts, alpha = .5)
#'
#' # qmplot is great for this kind of thing...
#' qmplot(lon, lat, data = chkpts, color = class, darken = .6)
#' qmplot(lon, lat, data = chkpts, geom = "density2d", color = class, darken = .6)
#'
#' ## maprange
#' ##################################################
#'
#' hdf <- get_map()
#' mu <- c(-95.3632715, 29.7632836)
#' points <- data.frame(MASS::mvrnorm(1000, mu = mu, diag(c(.1, .1))))
#' names(points) <- c("lon", "lat")
#' points$class <- sample(c("a","b"), 1000, replace = TRUE)
#'
#' ggmap(hdf) + geom_point(data = points) # maprange built into extent = panel, device
#' ggmap(hdf) + geom_point(aes(colour = class), data = points)
#'
#' ggmap(hdf, extent = "normal") + geom_point(data = points)
#' # note that the following is not the same as extent = panel
#' ggmap(hdf, extent = "normal", maprange = TRUE) + geom_point(data = points)
#'
#' # and if you need your data to run off on a extent = device (legend included)
#' ggmap(hdf, extent = "normal", maprange = TRUE) +
#'   geom_point(aes(colour = class), data = points) +
#'   theme_nothing(legend = TRUE) + theme(legend.position = "right")
#'
#' # again, qmplot is probably more useful
#' qmplot(lon, lat, data = points, color = class, darken = .4, alpha = I(.6))
#' qmplot(lon, lat, data = points, color = class, maptype = "toner-lite")
#'
#' ## cool examples
#' ##################################################
#'
#' # contour overlay
#' ggmap(get_map(maptype = "satellite"), extent = "device") +
#'   stat_density2d(aes(x = lon, y = lat, colour = class), data = chkpts, bins = 5)
#'
#'
#' # adding additional content
#' library(grid)
#' baylor <- get_map("one bear place, waco, texas", zoom = 15, maptype = "satellite")
#' ggmap(baylor)
#'
#' # use gglocator to find lon/lat"s of interest
#' (clicks <- gglocator(2) )
#' ggmap(baylor) +
#'   geom_point(aes(x = lon, y = lat), data = clicks, colour = "red", alpha = .5)
#' expand.grid(lon = clicks$lon, lat = clicks$lat)
#'
#' ggmap(baylor) + theme_bw() +
#'   annotate("segment", x=-97.110, xend=-97.1188, y=31.5450, yend=31.5485,
#'     colour=I("red"), arrow = arrow(length=unit(0.3,"cm")), size = 1.5) +
#'   annotate("label", x=-97.113, y=31.5445, label = "Department of Statistical Science",
#'     colour = I("red"), size = 3.5) +
#'   labs(x = "Longitude", y = "Latitude") + ggtitle("Baylor University")
#'
#'
#' baylor <- get_map("marrs mclean science, waco, texas", zoom = 16, maptype = "satellite")
#'
#' ggmap(baylor, extent = "panel") +
#'   annotate("segment", x=-97.1175, xend=-97.1188, y=31.5449, yend=31.5485,
#'     colour=I("red"), arrow = arrow(length=unit(0.4,"cm")), size = 1.5) +
#'   annotate("label", x=-97.1175, y=31.5447, label = "Department of Statistical Science",
#'     colour = I("red"), size = 4)
#'
#'
#'
#' # a shapefile like layer
#' data(zips)
#' ggmap(get_map(maptype = "satellite", zoom = 8), extent = "device") +
#'   geom_polygon(aes(x = lon, y = lat, group = plotOrder),
#'     data = zips, colour = NA, fill = "red", alpha = .2) +
#'   geom_path(aes(x = lon, y = lat, group = plotOrder),
#'     data = zips, colour = "white", alpha = .4, size = .4)
#'
#' library(plyr)
#' zipsLabels <- ddply(zips, .(zip), function(df){
#'   df[1,c("area", "perimeter", "zip", "lonCent", "latCent")]
#' })
#' ggmap(get_map(maptype = "satellite", zoom = 9),
#'     extent = "device", legend = "none", darken = .5) +
#'   geom_text(aes(x = lonCent, y = latCent, label = zip, size = area),
#'     data = zipsLabels, colour = I("red")) +
#'   scale_size(range = c(1.5,6))
#'
#' qmplot(lonCent, latCent, data = zipsLabels, geom = "text",
#'   label = zip, size = area, maptype = "toner-lite", color = I("red")
#' )
#'
#'
#' ## crime data example
#' ##################################################
#'
#' # only violent crimes
#' violent_crimes <- subset(crime,
#'   offense != "auto theft" &
#'   offense != "theft" &
#'   offense != "burglary"
#' )
#'
#' # rank violent crimes
#' violent_crimes$offense <-
#'   factor(violent_crimes$offense,
#'     levels = c("robbery", "aggravated assault",
#'       "rape", "murder")
#'   )
#'
#' # restrict to downtown
#' violent_crimes <- subset(violent_crimes,
#'   -95.39681 <= lon & lon <= -95.34188 &
#'    29.73631 <= lat & lat <=  29.78400
#' )
#'
#'
#' # get map and bounding box
#' theme_set(theme_bw(16))
#' HoustonMap <- qmap("houston", zoom = 14, color = "bw",
#'   extent = "device", legend = "topleft")
#' HoustonMap <- ggmap(
#'   get_map("houston", zoom = 14, color = "bw"),
#'   extent = "device", legend = "topleft"
#' )
#'
#' # the bubble chart
#' HoustonMap +
#'    geom_point(aes(x = lon, y = lat, colour = offense, size = offense), data = violent_crimes) +
#'    scale_colour_discrete("Offense", labels = c("Robbery","Aggravated Assault","Rape","Murder")) +
#'    scale_size_discrete("Offense", labels = c("Robbery","Aggravated Assault","Rape","Murder"),
#'      range = c(1.75,6)) +
#'    guides(size = guide_legend(override.aes = list(size = 6))) +
#'    theme(
#'      legend.key.size = grid::unit(1.8,"lines"),
#'      legend.title = element_text(size = 16, face = "bold"),
#'      legend.text = element_text(size = 14)
#'    ) +
#'    labs(colour = "Offense", size = "Offense")
#'
#'
#' # doing it with qmplot is even easier
#' qmplot(lon, lat, data = violent_crimes, maptype = "toner-lite",
#'   color = offense, size = offense, legend = "topleft"
#' )
#'
#' # or, with styling:
#' qmplot(lon, lat, data = violent_crimes, maptype = "toner-lite",
#'   color = offense, size = offense, legend = "topleft"
#' ) +
#'   scale_colour_discrete("Offense", labels = c("Robbery","Aggravated Assault","Rape","Murder")) +
#'   scale_size_discrete("Offense", labels = c("Robbery","Aggravated Assault","Rape","Murder"),
#'     range = c(1.75,6)) +
#'   guides(size = guide_legend(override.aes = list(size = 6))) +
#'   theme(
#'     legend.key.size = grid::unit(1.8,"lines"),
#'     legend.title = element_text(size = 16, face = "bold"),
#'     legend.text = element_text(size = 14)
#'   ) +
#'   labs(colour = "Offense", size = "Offense")
#'
#'
#'
#'
#'
#'
#' # a contour plot
#' HoustonMap +
#'   stat_density2d(aes(x = lon, y = lat, colour = offense),
#'     size = 3, bins = 2, alpha = 3/4, data = violent_crimes) +
#'    scale_colour_discrete("Offense", labels = c("Robbery","Aggravated Assault","Rape","Murder")) +
#'    theme(
#'      legend.text = element_text(size = 15, vjust = .5),
#'      legend.title = element_text(size = 15,face="bold"),
#'      legend.key.size = grid::unit(1.8,"lines")
#'    )
#'
#'
#'
#' # 2d histogram...
#' HoustonMap +
#'   stat_bin_2d(aes(x = lon, y = lat, colour = offense, fill = offense),
#'     size = .5, bins = 30, alpha = 2/4, data = violent_crimes) +
#'    scale_colour_discrete("Offense",
#'      labels = c("Robbery","Aggravated Assault","Rape","Murder"),
#'      guide = FALSE) +
#'    scale_fill_discrete("Offense", labels = c("Robbery","Aggravated Assault","Rape","Murder")) +
#'    theme(
#'      legend.text = element_text(size = 15, vjust = .5),
#'      legend.title = element_text(size = 15,face="bold"),
#'      legend.key.size = grid::unit(1.8,"lines")
#'    )
#'
#'
#'
#'
#'
#' # changing gears (get a color map)
#' houston <- get_map("houston", zoom = 14)
#' HoustonMap <- ggmap(houston, extent = "device", legend = "topleft")
#'
#' # a filled contour plot...
#' HoustonMap +
#'   stat_density2d(aes(x = lon, y = lat, fill = ..level.., alpha = ..level..),
#'     size = 2, bins = 4, data = violent_crimes, geom = "polygon") +
#'   scale_fill_gradient("Violent\nCrime\nDensity") +
#'   scale_alpha(range = c(.4, .75), guide = FALSE) +
#'   guides(fill = guide_colorbar(barwidth = 1.5, barheight = 10))
#'
#' # ... with an insert
#'
#' overlay <- stat_density2d(aes(x = lon, y = lat, fill = ..level.., alpha = ..level..),
#'     bins = 4, geom = "polygon", data = violent_crimes)
#'
#' attr(houston,"bb") # to help finding (x/y)(min/max) vals below
#'
#' HoustonMap +
#'   stat_density2d(aes(x = lon, y = lat, fill = ..level.., alpha = ..level..),
#'     bins = 4, geom = "polygon", data = violent_crimes) +
#'   scale_fill_gradient("Violent\nCrime\nDensity") +
#'   scale_alpha(range = c(.4, .75), guide = FALSE) +
#'   guides(fill = guide_colorbar(barwidth = 1.5, barheight = 10)) +
#'   inset(
#'     grob = ggplotGrob(ggplot() + overlay +
#'       scale_fill_gradient("Violent\nCrime\nDensity") +
#'       scale_alpha(range = c(.4, .75), guide = FALSE) +
#'       theme_inset()
#'     ),
#'     xmin = -95.35877, xmax = -95.34229,
#'     ymin = 29.73754, ymax = 29.75185
#'   )
#'
#'
#'
#'
#'
#'
#'
#'
#'
#' ## more examples
#' ##################################################
#'
#' # you can layer anything on top of the maps (even meaningless stuff)
#' df <- data.frame(
#'   lon = rep(seq(-95.39, -95.35, length.out = 8), each = 20),
#'   lat = sapply(
#'     rep(seq(29.74, 29.78, length.out = 8), each = 20),
#'     function(x) rnorm(1, x, .002)
#'   ),
#'   class = rep(letters[1:8], each = 20)
#' )
#'
#' qplot(lon, lat, data = df, geom = "boxplot", fill = class)
#'
#' HoustonMap +
#'   geom_boxplot(aes(x = lon, y = lat, fill = class), data = df)
#'
#'
#'
#'
#' ## the base_layer argument - faceting
#' ##################################################
#'
#' df <- data.frame(
#'   x = rnorm(1000, -95.36258, .2),
#'   y = rnorm(1000,  29.76196, .2)
#' )
#'
#' # no apparent change because ggmap sets maprange = TRUE with extent = "panel"
#' ggmap(get_map(), base_layer = ggplot(aes(x = x, y = y), data = df)) +
#'   geom_point(colour = "red")
#'
#' # ... but there is a difference
#' ggmap(get_map(), base_layer = ggplot(aes(x = x, y = y), data = df), extent = "normal") +
#'   geom_point(colour = "red")
#'
#' # maprange can fix it (so can extent = "panel")
#' ggmap(get_map(), maprange = TRUE, extent = "normal",
#'   base_layer = ggplot(aes(x = x, y = y), data = df)) +
#'   geom_point(colour = "red")
#'
#' # base_layer makes faceting possible
#' df <- data.frame(
#'   x = rnorm(10*100, -95.36258, .075),
#'   y = rnorm(10*100,  29.76196, .075),
#'   year = rep(paste("year",format(1:10)), each = 100)
#' )
#' ggmap(get_map(), base_layer = ggplot(aes(x = x, y = y), data = df)) +
#'   geom_point() +  facet_wrap(~ year)
#'
#' ggmap(get_map(), base_layer = ggplot(aes(x = x, y = y), data = df), extent = "device") +
#'   geom_point() +  facet_wrap(~ year)
#'
#' qmplot(x, y, data = df)
#' qmplot(x, y, data = df, facets = ~ year)
#'
#'
#' ## neat faceting examples
#' ##################################################
#'
#' # simulated example
#' df <- data.frame(
#'   x = rnorm(10*100, -95.36258, .05),
#'   y = rnorm(10*100,  29.76196, .05),
#'   year = rep(paste("year",format(1:10)), each = 100)
#' )
#' for(k in 0:9){
#'   df$x[1:100 + 100*k] <- df$x[1:100 + 100*k] + sqrt(.05)*cos(2*pi*k/10)
#'   df$y[1:100 + 100*k] <- df$y[1:100 + 100*k] + sqrt(.05)*sin(2*pi*k/10)
#' }
#'
#' ggmap(get_map(),
#'   base_layer = ggplot(aes(x = x, y = y), data = df)) +
#'   stat_density2d(aes(fill = ..level.., alpha = ..level..),
#'     bins = 4, geom = "polygon") +
#'   scale_fill_gradient2(low = "white", mid = "orange", high = "red", midpoint = 10) +
#'   scale_alpha(range = c(.2, .75), guide = FALSE) +
#'   facet_wrap(~ year)
#'
#'
#'
#' # crime example by month
#' levels(violent_crimes$month) <- paste(
#'   toupper(substr(levels(violent_crimes$month),1,1)),
#'   substr(levels(violent_crimes$month),2,20), sep = ""
#' )
#' houston <- get_map(location = "houston", zoom = 14, source = "osm", color = "bw")
#' HoustonMap <- ggmap(houston,
#'   base_layer = ggplot(aes(x = lon, y = lat), data = violent_crimes)
#'   )
#'
#' HoustonMap +
#'   stat_density2d(aes(x = lon, y = lat, fill = ..level.., alpha = ..level..),
#'     bins = I(5), geom = "polygon", data = violent_crimes) +
#'   scale_fill_gradient2("Violent\nCrime\nDensity",
#'     low = "white", mid = "orange", high = "red", midpoint = 500) +
#'   labs(x = "Longitude", y = "Latitude") + facet_wrap(~ month) +
#'   scale_alpha(range = c(.2, .55), guide = FALSE) +
#'   ggtitle("Violent Crime Contour Map of Downtown Houston by Month") +
#'   guides(fill = guide_colorbar(barwidth = 1.5, barheight = 10))
#'
#'
#'
#'
#'
#'
#'
#'
#' ## darken argument
#' ##################################################
#' ggmap(get_map())
#' ggmap(get_map(), darken = .5)
#' ggmap(get_map(), darken = c(.5,"white"))
#' ggmap(get_map(), darken = c(.5,"red")) # silly, but possible
#'
#'
#' }
ggmap <- function(ggmap, extent = "panel", base_layer, maprange = FALSE,
  legend = "right", padding = .02, darken = c(0, "black"), b, fullpage, expand, ...)
{

  # dummies to trick R CMD check
  lon <- NULL; rm(lon); lat <- NULL; rm(lat); fill <- NULL; rm(fill);
  ll.lon <- NULL; rm(ll.lon); ur.lon <- NULL; rm(ur.lon);
  ll.lat <- NULL; rm(ll.lat); ur.lat <- NULL; rm(ur.lat);

  # deprecated syntaxes
  if(!missing(b)) {
    .Deprecated(msg = "b syntax deprecated, use padding.")
    padding <- b
  }

  if(!missing(fullpage) || !missing(expand)){
    .Deprecated(msg = "fullpage and expand syntaxes deprecated, use extent.")
    if(missing(fullpage)) { fullpage <- FALSE }
    if(missing(expand)) { expand <- FALSE }
    if(fullpage) extent <- "device"
    if(fullpage == FALSE && expand == TRUE) extent <- "panel"
    if(fullpage == FALSE && expand == FALSE) extent <- "normal"
  }



  # check arguments
  if(class(ggmap)[1] != "ggmap"){
    stop("ggmap plots objects of class ggmap, see ?get_map", call. = FALSE)
  }

  match.arg(legend, c("right", "left", "bottom", "top",
    "bottomleft", "bottomright", "topleft", "topright", "none"))

  if(is.language(darken)) darken <- eval(darken) # happens when passed from qmap


  # check darken
  stopifnot(0 <= as.numeric(darken[1]) && as.numeric(darken[1]) <= 1)
  if(length(darken) == 1 & is.numeric(darken)) darken <- c(darken, "black")


  # make raster plot or tile plot
  if(missing(base_layer) || base_layer == "auto"){
    if(inherits(ggmap, "raster")){ # raster
      # make base layer data.frame
      fourCorners <- expand.grid(
    	  lon = as.numeric(attr(ggmap, "bb")[,c("ll.lon","ur.lon")]),
  	    lat = as.numeric(attr(ggmap, "bb")[,c("ll.lat","ur.lat")])
      )

      # shorthand notation
      xmin <- attr(ggmap, "bb")$ll.lon
      xmax <- attr(ggmap, "bb")$ur.lon
      ymin <- attr(ggmap, "bb")$ll.lat
  	  ymax <- attr(ggmap, "bb")$ur.lat

      p <- ggplot(aes(x = lon, y = lat), data = fourCorners) +
  	    geom_blank() +
  	    inset_raster(ggmap, xmin, xmax, ymin, ymax) +
  	    annotate("rect", xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax,
  	      fill = darken[2], alpha = as.numeric(darken[1]))

    } else { # tile, depricated
      p <- ggplot() + geom_tile(aes(x = lon, y = lat, fill = fill), data = ggmap) +
        scale_fill_identity(guide = "none")
      .Deprecated(msg = "geom_tile method is deprecated, use rasters.")
    }
  } else { # base_layer provided making facets possible
    # get call
  	stopifnot(inherits(ggmap, "raster"))
    args <- as.list(match.call()[-1])
    base <- deparse(args$base_layer) # "ggplot(aes(), data = blah)"
    # if passed from another function
    if(base[1] == "base_layer") base <- deparse(eval(args$base_layer))

    # shorthand notation
    xmin <- attr(ggmap, "bb")$ll.lon
    xmax <- attr(ggmap, "bb")$ur.lon
  	ymin <- attr(ggmap, "bb")$ll.lat
  	ymax <- attr(ggmap, "bb")$ur.lat
    str2parse <- paste(paste(base, collapse = ""), "geom_blank()",
      "inset_raster(ggmap, xmin, xmax, ymin, ymax)",
      sep = " + "
    )

    # attempt to ensure that base_layer argument is evaluated correctly by forcing evaluation in the calling environment.
    # if that fails, the previous default behavior is provided. see https://github.com/dkahle/ggmap/issues/84 for background.
    p <- try(eval(parse(text = paste(base,collapse="")),parent.frame(1))+geom_blank()+inset_raster(ggmap, xmin, xmax, ymin, ymax))
    if(inherits(p,"try-error")) p <- eval(parse(text = str2parse))

    p <- p + annotate("rect", xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax,
  	  fill = darken[2], alpha = as.numeric(darken[1]))
  }

  # enforce maprange
  if(maprange) p <- p + xlim(xmin, xmax) + ylim(ymin, ymax)

  # set scales
  p <- p + coord_map2()

  # set extent
  xmin <- attr(ggmap, "bb")$ll.lon
  xmax <- attr(ggmap, "bb")$ur.lon
  ymin <- attr(ggmap, "bb")$ll.lat
  ymax <- attr(ggmap, "bb")$ur.lat

  if(extent == "normal"){
    # nothing
  } else if(extent == "panel"){
  	p <- p +
      scale_x_continuous(limits = c(xmin, xmax), expand = c(0,0)) +
      scale_y_continuous(limits = c(ymin, ymax), expand = c(0,0))
  } else if(extent == "device"){
  	p <- p +
      scale_x_continuous(limits = c(xmin, xmax), expand = c(0,0)) +
      scale_y_continuous(limits = c(ymin, ymax), expand = c(0,0)) +
      theme_nothing(legend = TRUE)

    # legend for full device map
    if(legend %in% c("topleft","topright","bottomleft","bottomright")){
      if(legend == "bottomleft"){
        lp <- c(padding, padding)
        lj <- c(0,0)
      } else if(legend == "topleft"){
        lp <- c(padding, 1-padding)
        lj <- c(0,1)
      } else if(legend == "bottomright"){
        lp <- c(1-padding, padding)
        lj <- c(1,0)
      } else if(legend == "topright"){
        lp <- c(1-padding, 1-padding)
        lj <- c(1,1)
      }
      p <- p + theme(
        legend.position = lp, legend.justification = lj,
        legend.background = element_rect(
          fill = "white", colour = "gray80", size = .2
        )
      )
    } else if(legend %in% c("left","right","bottom","top")){
      p <- p + theme(legend.position = legend)
    } # else legend = "none" as part of theme_nothing()
  }

  # return plot
  p
}





#' Don't use this function, use ggmap.
#'
#' ggmap plots the raster object produced by \code{\link{get_map}}.
#'
#' @param ggmap an object of class ggmap (from function get_map)
#' @param fullpage logical; should the map take up the entire viewport?
#' @param base_layer a ggplot(aes(...), ...) call; see examples
#' @param maprange logical for use with base_layer; should the map define the x and y limits?
#' @param expand should the map extend to the edge of the panel? used with base_layer and maprange=TRUE.
#' @param ... ...
#' @return a ggplot object
#' @author David Kahle \email{david.kahle@@gmail.com}
#' @seealso \code{\link{get_map}}, \code{\link{qmap}}
#' @export
#' @examples
#' \dontrun{
#' this is a deprecated function, use ggmap.
#' }
ggmapplot <- function(ggmap, fullpage = FALSE,
  base_layer, maprange = FALSE, expand = FALSE, ...)
{
  .Deprecated(msg = "ggmapplot syntax deprecated, use ggmap.")
  ggmap(ggmap, fullpage = fullpage, base_layer = base_layer,
    maprange = FALSE, expand = FALSE, ggmapplot = TRUE)
}
